MoltenVK/MoltenVK/GPUObjects/MVKQueryPool.mm

```diff
@@ -24,9 +24,112 @@
 #include "MVKOSExtensions.h"
 #include "MVKFoundation.h"
 #include <sys/mman.h>
+#include <mach/mach_time.h>
+#include <inttypes.h>
 
 using namespace std;
 
+/** Converts a timestamp in GPU time into a timestamp in CPU time in nanoseconds, using the previously recorded timestamp correlation marker. */
+static MTLTimestamp adjustGPUTime(const TimestampCorrelationMarker& timestampCorrelationMarker, MTLTimestamp gpuTime)
+{
+	// Time in [0,1] domain = ( x - a ) / ( b - a )
+	double normalizedGpuTime = (gpuTime - timestampCorrelationMarker.gpuStart) / (double)(timestampCorrelationMarker.gpuEnd - timestampCorrelationMarker.gpuStart);
+	// Take the time in the [0,1] domain and scale it by the CPU
+	// lapse, then add the CPU start time to match the CPU’s timestamp.
+	double adjGpuTimeStamp = timestampCorrelationMarker.cpuStart + (timestampCorrelationMarker.cpuEnd - timestampCorrelationMarker.cpuStart) * normalizedGpuTime;
+	
+	// Convert from CPU counter to nanoseconds
+	mach_timebase_info_data_t time_info;
+	mach_timebase_info(&time_info);
+
+	return adjGpuTimeStamp * time_info.numer / time_info.denom;
+}
+
+constexpr uint32_t MVKTimestampBuffers::kTimestampCountPerBuffer;
+
+MVKTimestampBuffers::MVKTimestampBuffers(MVKDevice* device)
+	:_device(device), _nextTimestampIndex(0), _isResolved(false)
+{
+	
+}
+
+uint32_t MVKTimestampBuffers::allocateTimestamp(id<MTLCounterSampleBuffer>& outSampleBuffer)
+{
+	const uint32_t bufferCount = (uint32_t)_bufferPool.size();
+	const uint32_t destinationBufferIndex = _nextTimestampIndex / kTimestampCountPerBuffer;
+	
+	if(destinationBufferIndex == bufferCount)
+	{
+		__autoreleasing NSError* error = nil;
+		MTLCounterSampleBufferDescriptor* timestampSamplerDesc = [[MTLCounterSampleBufferDescriptor alloc] init];
+		timestampSamplerDesc.counterSet = _device->getTimestampCounterSet();
+		timestampSamplerDesc.storageMode = MTLStorageModeShared;
+		timestampSamplerDesc.sampleCount = kTimestampCountPerBuffer;
+		
+		id<MTLCounterSampleBuffer> newCounterSampleBuffer = [_device->getMTLDevice() newCounterSampleBufferWithDescriptor:timestampSamplerDesc error:&error];
+		
+		if(error != nil) {
+			_device->reportError(VK_ERROR_INITIALIZATION_FAILED,
+					"Failed to initialize timestamp sample buffer (Error code %li):\n%s",
+					(long)error.code, error.localizedDescription.UTF8String);
+		}
+		
+		_bufferPool.push_back(newCounterSampleBuffer);
+	}
+	
+	outSampleBuffer = _bufferPool[destinationBufferIndex];
+	
+	_isResolved = false;
+	const uint32_t queryIndex = _nextTimestampIndex++;
+	return queryIndex % kTimestampCountPerBuffer;
+}
+
```

Comment:
Do we really need a buffer per query? I was thinking we'd do a buffer per pool, and each timestamp query would correspond to a sample location. Vulkan doesn't allow writing a query multiple times anyway.

Reply:
This buffer is not per queries, but for timing samples. The queries themselves are kept as a separate concept that take from the sample buffer as needed.

Doing this a per-query pool basis was my initial approach, but ultimately I had to separate this into different concepts due to the encoder issue mentioned in my other comment.

Reply:
Also note we're not allocating a buffer per query, but rather a buffer per `kTimestampCountPerBuffer` queries.